File: src/scala/22.scala
/*
[minhuang@mocha:scala]$ time scala problem-00022.scala 
871198282

real  0m2.015s
user  0m0.988s
sys 0m0.084s
*/
import scala.io.Source
import scala.Math._

val scores = Map() ++ (('A' to 'Z').toList zip (1 to 26).toList)

def score(name: String) = name.map(c => scores(c)).reduceLeft(_ + _)

val data = Source.fromFile("../../data/22.txt")
                 .getLines
                 .toList
                 .first
                 .replaceAll("\"", "")
                 .split(",")
                 .toList

def f(t: Pair[Int, String]) = (t._1 + 1) * score(t._2)

val result = (data.indices zip data.sort(_ < _)).map(f _).reduceLeft(_ + _)
println(result)

File: ruby/22.rb
def score(name)
  scores = Hash[('A'..'Z').zip(1..26)]
  name.chars.map { |c| scores[c] }.inject(&:+)
end

file = File.new("../data/22.txt")
data = file.gets.gsub("\"", "").split(",") # All one line in file
file.close

result = data.each_index.
              zip(data.sort).
              map { |t| (t.first+1) * score(t.last) }.
              inject(&:+)
puts result
